updates to Doxygen documentation, including Release Notes and User Guide

jbathegit opened the pull request:
Fixes #69  
Fixes #74 

Part of #48 - the URL for the external HTML documentation is now in the new UserGuide.md

edwardhartnett commented on docs/Mainpage.md:
```diff
@@ -0,0 +1,8 @@
+
```
My goal was for all the doxygen docs to use the same file name: user_guide.md.

So can we keep that name and keep that as our main page for this repo as well?

jbathegit replied:
Yes this is doable, but it will be misleading because now the main page would be called user_guide.md, and there would be a separate md file which now contains the actual user guide, so what should we call that?  In other words, right now we have the following in nceplibs-bufr/docs:

`Mainpage.md` - the main page with overall library description
`ReleaseNotes.md` - the release notes for each version
`UserGuide.md` - the actual user guide; right now this is just a link to the HTML docs but will eventually contain the reformulated markdown content

So again, if we now rename Mainpage.md as user_guide.md, then what do we call UserGuide.md?  I would actually argue that it makes more sense to have a common Mainpage.md (or we can call it main_page.md or whatever you want) in every repo, and if you want we can also require a separate release notes file and separate user guide file.  But the point is I don't think we should dictate that the user guide has to be the main page in every repo.

My $0.02 anyway.

edwardhartnett replied:
I think we should dictate that the user_guide.md is the main page. I don't see any benefit of having different names in different project. ;-)

So have a file called user_guide.md which is the doxygen mainpage.

I set them all up the same way to make it easy to work on different projects.

jbathegit replied:
Understood, and again I can do that, but the question still remains: if I want to have the actual User Guide documentation in a different file, then what should I call that file if the name `user_guide.md` is already taken as the name for the main page?  Suggestions?  I can certainly make up any filename I want, but is this something you'd also want to standardize across all projects?

It will just be a bit confusing to have the main page named as `user_guide.md` when the User Guide itself is located elsewhere in a file with a different name.

edwardhartnett commented on docs/UserGuide.md:
```diff
@@ -0,0 +1,4 @@
+
```
Please don't change the file name, other repos also use user_guide.md and it would be good to have them all the same, for ease of working on different projects.

jbathegit replied:
see above ;-)

edwardhartnett commented on src/adn30.f:
```diff
@@ -1,37 +1,35 @@
-C> this routine is called by: cadn30() dxinit() igetrfel() istdesc()
+C> <b>This routine calls:</b> bort()
+C>
+C> <b>this routine is called by:</b> cadn30() dxinit() igetrfel() istdesc()
```
I don't see a </b> to match this bold tag...

jbathegit replied:
Not sure where you're looking, but I do see a matching </b> for each </b> in both lines 27 and 29.

edwardhartnett commented on src/bvers.f.in:
```diff
@@ -1,13 +1,16 @@
 
-C> @brief This subroutine returns a character string containing
-C> the version number of the BUFR archive library software.
+C> @author J. Ator
```
I think the brief sentance for the subroutine has to come first.

jbathegit replied:
OK, I'll fix this.

edwardhartnett replied:
Also I believe we found that we don't need the brief tag in the subroutine description, but we do need it in the file description.

jbathegit replied:
Agreed, and right now it's only in the file block.  You may just be confused by the + and - labels in the diff output.

jbathegit commented:
OK, I just went ahead and renamed Mainpage.md back to user_guide.md.  So that will still be the main page, but we'll just have a separate UserGuide.md that will contain the actual User Guide ;-)

I also updated the Release Notes as we discussed yesterday, including referencing the underlying GitHub issue numbers and using lowercase to automatically reference the subprograms in the Release Notes.

I also added `@brief` notes to UserGuide.md and ReleaseNotes.md as we discussed yesterday.  I found that as long as you put it after the filename (which is preceded by ####), then it works as we thought it might.

Finally, I went and looked through all of my notes, earlier issues and email archives, and I agree that we can go ahead and use 11.4.0 for the next version number, so I made those corresponding updates as well (including in Projects under GitHub).  If you want to see the net effect of what all of these latest updates looks like, see https://emc.ncep.noaa.gov/emc/pages/infrastructure/bufrlib_jeff/html/index.html

Hopefully you're OK now with the 2nd commit that I just pushed on top of the 1st one, but if not let me know.

jbathegit commented:
@edwardhartnett 

Just touching base - do you think you might have time to take a look at this and (hopefully? ;-) approve it soon?  I realize you're busy, so if you'd prefer I could ask Rahul or someone else to do this.  I'd just like to get these commits merged into the develop branch as soon as possible so I can in turn merge them into some other branches I'm working on.

In particular, I have an idea I want to test out for #26, but to do that I'll need to make another change to Doxyfile.in.inc, and I've already got some pending changes to that same file within this particular PR which are awaiting review.  I just don't want my various branches to get too out of whack and create a potential merge conflict down the road! ;-)

jbathegit commented:
Thanks Ed!

jbathegit commented:
@aerorahul @edwardhartnett @kgerheiser 

Something seems to have gone awry when merging my `jba_docs` branch back into `develop`   Please see https://github.com/NOAA-EMC/NCEPLIBS-bufr/actions/runs/373251691, then drill down to the install-deps step under the macos-10.15 build process.

I have no idea what this error means or where it's coming from as I've never seen this before, and certainly not during any of my earlier tests.  Would appreciate any help/insight/guidance - thanks!

jbathegit commented:
I see the following in the output from the install-deps step:

~~~
    running build_scripts
    creating build/scripts-3.9
    copying and adjusting bin/markdown_py -> build/scripts-3.9
    changing mode of build/scripts-3.9/markdown_py from 644 to 755
    running build_docs
    Converting docs/release-2.3.txt -> build/docs/release-2.3.html
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-install-vsbig6xl/markdown/setup.py", line 216, in <module>
        setup(
      File "/usr/local/Cellar/python@3.9/3.9.0_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/core.py", line 148, in 
~~~

but I have no idea what this means.  In fact, I don't even know what `docs/release-2.3.txt` is, as I've never seen a file by that name before!?

aerorahul commented:
Looks like python was updated in the runner to 3.9. 
Markdown which is a dependency of pdoc was not updated. 
We don't install markdown explicitly, it comes with pdoc. 

We can either figure out to use previous version of python 3.8.6 I think. 
Or wait for markdown to be fixed

jbathegit commented:
@aerorahul 

Sorry, I don't understand what you mean when you say 
> Or wait for markdown to be fixed

Just trying to understand what our options are here, b/c I'd really like to be able to resolve this so we can once again have error-free merging back into the `develop` branch ;-)

aerorahul commented:
The failure is installing the package markdown via `pip install`.
markdown 2.4.1 is being installed because it is a dependency to pdoc.
pdoc is how the documentation is being created for the python API.

The github runner updated the default python version from 3.8 to 3.9 as a result, all pip installs are now looking for packages built with 3.9.
It seems that the markdown dependency is not compatible with pdoc.

I am trying to figure out what our alternatives are.
One possibility is to turn off documentation creation in the CI. `-DENABLE_DOCS=OFF` till an update to the pdoc and markdown package is available on pypi.
It was turned ON in the CI with the PR that added the python API.
Another alternative is to wait till the update to pdoc and markdown
A third alternative is to find an alternative to pdoc for documentation such as doxypy

jbathegit commented:
FWIW, we're using Doxygen for all of our Fortran and C documentation for NCEPLIBS-bufr, as well as for most of the other documented content on the corresponding github.io site for NCEPLIBS-bufr.  Doxygen is also planned to one day be used in a similar manner for all other NCEPLIBS projects as well.

So you may(?) want to give some consideration to your third stated alternative (i.e. doxypy), because if we can all be using the same or similar tools for all of our documentation, then it would likely make future maintenance easier for everyone on the NCEPLIBS team, both within a given project as well as across different projects.

Just a thought ;-)
